circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/NavigationDecoration.kt

```diff
@@ -0,0 +1,91 @@
+// Copyright (C) 2022 Slack Technologies, LLC
+// SPDX-License-Identifier: Apache-2.0
+package com.slack.circuit.foundation
+
+import androidx.compose.animation.AnimatedContent
+import androidx.compose.animation.AnimatedContentScope
+import androidx.compose.animation.AnimatedContentTransitionScope
+import androidx.compose.animation.ContentTransform
+import androidx.compose.animation.ExperimentalSharedTransitionApi
+import androidx.compose.animation.core.Transition
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.Stable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import com.slack.circuit.backstack.NavArgument
+import com.slack.circuit.backstack.NavDecoration
+import com.slack.circuit.runtime.screen.Screen
+import com.slack.circuit.sharedelements.ProvideAnimatedTransitionScope
+import com.slack.circuit.sharedelements.SharedElementTransitionScope.AnimatedScope.Navigation
+import kotlinx.collections.immutable.ImmutableList
+
+@OptIn(ExperimentalSharedTransitionApi::class)
+public abstract class AnimatedNavDecoration(
+  private val decoratorFactory: AnimatedNavDecorator.Factory
+) : NavDecoration {
+
+  @Composable
+  public override fun <T : NavArgument> DecoratedContent(
+    args: ImmutableList<T>,
+    backStackDepth: Int,
+    modifier: Modifier,
+    content: @Composable (T) -> Unit,
+  ) {
+    val decorator = remember {
+      @Suppress("UNCHECKED_CAST")
+      decoratorFactory.create<T>() as AnimatedNavDecorator<T, AnimatedNavState>
+    }
+    with(decorator) {
+      Content(args, backStackDepth, modifier) { modifier ->
+        AnimatedContent(modifier = modifier, transitionSpec = transitionSpec()) { targetState ->
+          ProvideAnimatedTransitionScope(Navigation, this) {
+            AnimatedNavContent(targetState) { content(it) }
```

Comment:
it's hard for me to look at this level of nesting and feel like these new compose animation APIs are supposed to be easy :(